### DISCUSSION TITLE
Beta version of virtual temperature feature

### DISCUSSION THREAD

**Discussion by visualapproach** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#328] [Ideas]
Development branch - 2022-07-10

Hi community! Thought I'd say something about the latest update.
Often when stopping the pump and it's cold outside, the temp shown is lower than the actual temperature of the water body. This is due to the thermometer is located in the pump which cools down much quicker than the main water body. Now there is a virtual temperature, where the ESP tries to estimate the actual temperature, assuming it moves towards the ambient temperature. The virtual temp is corrected only when the pump is running and a new temp is reached by the real thermometer.
The table is self adjusting when pump is running with no heating.
There is default values that should be in the right ballpark, but to calibrate - Heat up to your preferred ready-temp (38-ish C). Then turn off the heater but keep the pump going. Every temperature change on the way towards ambient temperature, it will adjust and save the cooling rate. So next time it will be quite accurate. Or at least that's the goal. Since we now have a higher resolution thermometer, that takes the ambient temperature into account, we can make a better estimate of how long it will take to reach target temp. The old method "time to target temp" is still there in addition to this "time to ready".
I don't have any tub to test it on for the moment, so you are welcome to try it out and give feedback here.
By the way - I also made three channels of temperature to web sockets and mqtt - Celsius, Farenheit and whatever the pump is set to. HA sensors should now not be so confused about the unit. You may have to remove the old device in HA and restart.
I updated the YAML examples with an example to get HA to update the ambient temperature to the ESP.
![image](https://user-images.githubusercontent.com/22987250/178160862-4cd72cab-9cdb-4135-80ac-8f77da23884d.png)

**Comment by chunkysteveo** [comment]
Worth noting (not sure what/how you're doing the maths?!) that the difference can go the other way - for me, when the pump is off, and the Sun is out, the temp read by the unit is WAY higher than the water temperature, like a "greenhouse effect" I think? The sensor in the tubes gets hotter, and turning on the pump flushes out the static water and the temperature value drops to the actual water temp.

Not sure if that's helpful??

Would be good also to have some pre-built code with a flag to enable/disable an external temperature probe, and to hook up to one of the spare pins on the PCB, either a Dallas probe or a DHT11/22 sensor. People can choose if that's an external temp, case internal temp, throw it in the water extra temp, etc. Default disabled as most won't have or want that extra sensor.

**Reply by visualapproach** [reply]
Thank you! The temperature readings is discarded when the pump is off so it shouldn't matter if it hotter or colder. When pump is off it ticks at a certain rate towards the ambient temperature. The rate is increasing the further away from ambient you are. Saved in file "coolarray.json".
But it's great with these comments so I can handle situations I haven't thought of.

**Reply by visualapproach** [reply]
And one more thing. I tried to connect a temp probe (Dallas) to D0/D8 but it's a no go for different reasons. Maybe if we move the audio pin but not sure.

**Comment by Bishbosh02** [comment]
I added this to my system value_template: "{{ value_json.VTM | round(2)}}" as only want the temp to 2 decimal places, also I had a blip when the pump started and I turned the heater on
![TempBlip](https://user-images.githubusercontent.com/9953147/178531435-3594c64a-e83f-4b4b-b233-4026e3656e45.JPG)

**Reply by visualapproach** [reply]
Do you know how much effort I put in to get to that last decimal?! 😀 
Thanks for testing! Did the real temperature sensor also spike? And what's the vertical scale here?Hard to tell if it's one degree or 10.

**Reply by Bishbosh02** [reply]
The blip is at 10:01 when the heater is turned on Pump is turned on at 10:00, I will test it with the heater coming on after the pump by 30mins to see if we get the same blip tomorrow (hot tub temp is the actual water temp I have a ESPHome linked to a Dallas probe )

Ambient was the default 20deg C, have just added the automation to set the ambient from a blink camera  that is out side

![VtempBlip2](https://user-images.githubusercontent.com/9953147/178562460-3956fb1d-ceba-4e37-a6a6-b99c7bd8a78a.JPG)

**Reply by Bishbosh02** [reply]
No blip this time just a 2 deg offset between the two temps but the drop follows the real temp very close 👍 

![VtempBlip3](https://user-images.githubusercontent.com/9953147/178707862-b29436cf-0ea7-42a3-9850-55ab854230fd.JPG)

**Reply by visualapproach** [reply]
Thanks! I just uploaded a new version with bugfixes. The virtual temp should jump to "hot tub temp" after heating for a minute, but there was a bug. Also the self calibration had a few snags, but I don't think you run the pump without heat for cooling down so you probably didn't experience that.
Love HA graphs :)

**Reply by visualapproach** [reply]
BTW there is a new sensor "ambient temp" in HA with the new version. And rounded to 2 decimals :-)

**Reply by Bishbosh02** [reply]
OK Cool Built and installed

**Reply by Bishbosh02** [reply]
Note as we are having a mini heat wave in the UK, we are keeping at a steady 35deg C so that we can use it to cool down, at the weekend will probably change it to 30deg C

**Reply by Bishbosh02** [reply]
I also added this to node red
![NodeRedMod](https://user-images.githubusercontent.com/9953147/178772829-65d0b55a-8b2b-4a65-a2fe-45c3927fd8c3.JPG)

**Reply by Bishbosh02** [reply]
The new version has some odd blips

![VtempBlip4](https://user-images.githubusercontent.com/9953147/179347291-c88c44e1-26d3-4c3a-b167-904eee9f059d.JPG)

**Reply by visualapproach** [reply]
Doh! Could you add entity 'hot tub temp' to the graph also, please. Would make it easier to spot what's going on. Thx

**Reply by visualapproach** [reply]
And also the heater red stat

**Reply by Bishbosh02** [reply]
![VtempBlip5](https://user-images.githubusercontent.com/9953147/179350175-aca39844-35b6-4bbc-8254-335054788c73.JPG)

**Reply by Bishbosh02** [reply]
![VtempBlip6](https://user-images.githubusercontent.com/9953147/179356917-1db11b0d-9dd9-4367-a86c-539f46995d1a.JPG)

**Reply by visualapproach** [reply]
Thank you! I have fixed a couple of things now but not sure all bugs are out. Will upload to dev very soon. Also logging myself now after I disconnected the elements and pump and forcing sensor to read a flow to avoid E02 (I have no liner).

**Reply by Skeletorjus** [reply]
Probably not 100% related, but when the ESP8266 is restarted, the temperature is repoted as 0 across the board. Maybe it should hold off on publishing states over MQTT until the ESP has been running for a couple of seconds..?

**Reply by Bishbosh02** [reply]
One thing I found with the old version if you re-boot when you have an error (only way to clear the error) the unit setting changed to F so yes I think its best to wait after a re-boot before sending MQT info, I have just loaded the last dev vision 2022-07-16 and the readings look OK so-far....

**Reply by visualapproach** [reply]
Yes i'll fix the start up behavior.

**Reply by Bishbosh02** [reply]
![VtempBlip7](https://user-images.githubusercontent.com/9953147/179513198-4c111d92-d154-4adf-b9e0-10d4a62a4e24.JPG)

**Reply by visualapproach** [reply]
@Bishbosh02 weird. Can you add VTF sensor and AMB and zoom in the area from pump goes on? If you want.

**Reply by Bishbosh02** [reply]
![VtempBlip8](https://user-images.githubusercontent.com/9953147/179682194-d98557bc-b462-4e2d-98ac-b0751fb95d08.JPG)

**Comment by visualapproach** [comment]
Thank you. Can you verify that you had the 2022-07-16 version installed? I can't see why it should behave like that. It doesn't in my setup. Anyhow, I tweaked things a little again. Latest code 2022-07-19. I suggest you go to the web ui menu. Choose remove file. Enter coolarray.json and delete it. It will be created again with default values. It probably had some stupid values in it from the previous code.

**Reply by Bishbosh02** [reply]
Yes I was running 2022-07-16 and I have updated to 2022-07-19 and it looks good

**Reply by Bishbosh02** [reply]
I updated the code at 11:00 and the VT temp is just going up where spa temp and hot tub temp are constant (its a very hot day in the UK today 39deg C ) VT temp appears to try and track ambient when the pump is running ?

![VtempBlip9](https://user-images.githubusercontent.com/9953147/179800019-d0b56441-9635-4edb-9b9d-5250f430ce5e.JPG)

**Reply by visualapproach** [reply]
Yes, VT will go towards AMB at a rate saved in coolarray.json file. It stores 20 values- one for each degree off from ambient. So if your coolarray has too high numbers in it, it will go faster towards ambient than actual temperature does. Or if your tub is heavily insulated it will of course have a lower rate of change than the ESP thinks. Until it’s calibrated. Calibration is done when 1) pump is on and has been on for at least 5 min. 2) actual temp is changing a degree and no one messed with heater or bubbles, which would affect the rate we want to measure. 3) heater is off.
So, when hot tub temp raises one degree next time the VT will align to that and calculate the rate 1degree in x hours at this particular offset from ambient. And tomorrow the line will follow much closer, if all other things are the same, like difference between hot tub temp and ambient.

If you spot stupid behavior again you are welcome to tell me. Thx for your report

**Comment by visualapproach** [comment]
Yes, VT will go towards AMB at a rate saved in coolarray.json file. It stores 20 values- one for each degree off from ambient. So if your coolarray has too high numbers in it, it will go faster towards ambient than actual temperature does. Or if your tub is heavily insulated it will of course have a lower rate of change than the ESP thinks. Until it’s calibrated. Calibration is done when 1) pump is on and has been on for at least 5 min. 2) actual temp is changing a degree and no one messed with heater or bubbles, which would affect the rate we want to measure. 3) heater is off.
So, when hot tub temp raises one degree next time the VT will align to that and calculate the rate 1degree in x hours at this particular offset from ambient. And tomorrow the line will follow much closer, if all other things are the same, like difference between hot tub temp and ambient.

If you spot stupid behavior again you are welcome to tell me. Thx for your report

Hämta Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Bishbosh02 ***@***.***>
Skickat: Tuesday, July 19, 2022 6:23:12 PM
Till: visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA ***@***.***>
Kopia: visualapproach ***@***.***>; Author ***@***.***>
Ämne: Re: [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA] Beta version of virtual temperature feature (Discussion #328)


I updated the code at 11:00 and the VT temp is just going up where spa temp and hot tub temp are constant (its a very hot day in the UK today 39deg C ) VT temp appears to try and track ambient when the pump is running ?

[VtempBlip9]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F9953147%2F179800019-d0b56441-9635-4edb-9b9d-5250f430ce5e.JPG&data=05%7C01%7C%7C425349414cbf4fdf6c9508da69a301af%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637938446082870814%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=DsfTaqj%2F%2BbbRa4ah4I8VVX6C%2FCu2cA6w5x64EPu%2F6HY%3D&reserved=0>

**Reply by Bishbosh02** [reply]
OK thanks for the information I had just assumed that if the pump is running (VT temp = Measured Temp) and you only started using ambient temp to calculate VT temp when the pump was off

**Reply by visualapproach** [reply]
Yes that's a matter of choice. I chose to not follow the constant temperature because I thought it would be better to have fractional degrees. Good if it's good at estimating, otherwise not so good.

**Comment by GSzabados** [comment]
Any chance that ambient temperature can be update though MQTT or any other form, but not from the Web interface's slider?

**Reply by Skeletorjus** [reply]
This is possible, see this example:
https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/blob/master/Code/Home%20assistant%20config/update%20ambient%20temp%20automation%20snippet.yaml

**Reply by GSzabados** [reply]
Thanks! I haven't looked at that yet. Thanks for the pointer!

**Reply by GSzabados** [reply]
@Skeletorjus, thanks again! The Automation works well.

And here is a shout out to @visualapproach. The virtual temperature is top notch!

![Screenshot_20220928-161200_Home Assistant](https://user-images.githubusercontent.com/35445496/193230651-7d5d7eb6-ce89-4cd0-8554-5e276946c058.jpg)

**Reply by visualapproach** [reply]
Nice!

**Comment by visualapproach** [comment]
Development version moved to MASTER.
Development now includes a beta version of virtual temp for 4 wire systems. To calibrate, keep pump on, heater off until it has cooled to at least 2 C below setpoint. And don't use bubbles/jets while calibrating. Thx

**Reply by cobaltfish** [reply]
I'm trying to calibrate my 4 wire (Palm Springs Hydrojet) but I'm not sure its working.  Looking in the settings.txt "R" is currently 0.003768439, last updated at 13:16 (UTC?).  
I started the pump circulating at about 7:30 UTC, with set point, and water temp at 37C.  By about 13:45 UTC, actual water temp was down to 35C, in an ambient temperature that has ranged from about 8-15C.  
Within a moments of stopping the pump, the virtual temp was showing as the same as ambient temp.

What would you expect a typical R value to be? Based on normal usage, it looses about 0.3C per hour.  I see the initial default in Code/4-wire-version/lib/BWC4W/BWC_8266_4w.h is R_COOLING = 20 so less than 1 seems quite small

I diffed main.cpp between 4 and 6 wire, and uncommented the vtemp mqtt sections for a local build, so I can track it historically, rather than rely on direct monitoring, but have only just applied it, so have no graphs for the earlier run

**Reply by visualapproach** [reply]
@cobaltfish thank you for the detailed report! The 6- and 4-wire code is quite different. The inputs to changing states are all over the place in the 4-wire. Display and web and mqtt sends commands and the cio sends temperature. So it's harder to get right at first try. I'll give it another shot. Don't wait up tonight, I'm not sure I'll have time today.

**Reply by visualapproach** [reply]
Btw, @cobaltfish , I think what you describe will happen if you have spikes in the temperature readings. Do you?

**Reply by cobaltfish** [reply]
Thanks @visualapproach 
I had taken a look a few weeks back, having seen the virtual temp, wondering if it would be straightforward to port it to 4-wire, but didn't get so far as trying, so I'm very happy to test it for you :-)

I wasn't aware of any significant temp spikes, its been pretty steady all day.  
Looking at the graph since I started pulling in the vtemp via mqtt, its been doing huge (-1000) negative spikes!
<img width="1371" alt="image" src="https://user-images.githubusercontent.com/104502877/195156373-8d8405e7-d5ae-49db-ba6a-99e15bc28030.png">
(The smaller spike was when I ran the pump for a couple of mins, and then it only went down to about -1 before climbing to ambient of 12C)

**Reply by visualapproach** [reply]
Graphs are great debugging tool! I suspect those oscillations on the green curve ("currenttemperature") is the culprit. The device thinks the temp falls super quick and R goes bonkers. I think I can fix that.

**Reply by visualapproach** [reply]
Fix is uploaded: It will not calibrate if temperature reading changed the last half hour. So the wiggly green lines will be ignored after first edge.
Expect the virtual temperature to be nuts until next calibration is done, unless you edit the settings.txt file (by downloading, edit and uploading). 
It will only calibrate when currenttemperature differs more than 1 degree from ambient. Unfortunately it will also calibrate if you use airbubbles or jets in this experimental version. I want to see it work before putting more time in it.

**Reply by cobaltfish** [reply]
I think it did some mighty funny things when I first ran the pump after updating firmware.  Even though the pump was running, it was dropping down to the ambient temp, rather than tracking at about actual temp.
But I gave it a few mins, and now is tracking at 1 degree below actual:
<img width="1371" alt="image" src="https://user-images.githubusercontent.com/104502877/195190598-3a9691d1-3872-4659-99ad-9d78d32c68e2.png">
It should heat up off peak, so will see what its doing in the morning

(And thanks once again for the very speedy replies)

**Reply by cobaltfish** [reply]
so I ran the pump for 1 hour first thing, then stopped it, and it didn't work right:
<img width="1371" alt="image" src="https://user-images.githubusercontent.com/104502877/195288114-ebd9fadb-f0d4-4723-afa8-0597d89f8bc8.png">

To try and help with debugging, I've added a new mqtt output for /sensor/layzspa_4w_virtualtemp_r/config in main.cpp:
`  doc["dev"] = devicedoc["dev"];
  payload = "";
  topic = String(HA_PREFIX) + F("/sensor/layzspa_4w_virtualtemp_r/config");
  Serial.println(topic);
  doc["name"] = F("Layzspa virtual R");
  doc["uniq_id"] = "sensor.layzspa_4w_virtual_temp_R"+mychipid;
  doc["stat_t"] = mqttBaseTopic+F("/message");
  doc["val_tpl"] = F("{{ value_json.R | round(10) }}");
  doc["exp_aft"] = 700;
  doc["avty_t"] = mqttBaseTopic+F("/Status");
  doc["pl_avail"] = F("Alive");
  doc["pl_not_avail"] = F("Dead");
  if (serializeJson(doc, payload) == 0)
  {
    Serial.println(F("Failed to serialize HA message!"));
    return;
  }
  mqttClient.publish(topic.c_str(), payload.c_str(), true);
  mqttClient.loop();
  Serial.println(payload);
  doc.clear();
  doc.garbageCollect();`
  and added
  doc["R"] = R_COOLING; in BWC_8266_4w.cpp in the tring BWC::getJSONStates() section.  

If I've understood the code right, that should pull the current / latest value, rather than the one in settings.txt

**Reply by visualapproach** [reply]
Yes that's right, the R-value is loaded at startup and then updated continuously when the states are right. 
I don't think 1 hour is enough, you need to keep the pump on and nothing else until it registers a complete period, say from 37 to 35 degrees. It can then measure how long it took from switching to 36 to switching to 35.

**Reply by cobaltfish** [reply]
heh, I thought to myself just after I posted "perhaps I should have left it running longer"
Patience is a virtue!
I started the pump at 10 am, at about 11, it dropped 1c, and calculated R =37
At about 13:50, it dropped to 33 and recalculated R = 57

Just as I started to post this and get screen grabs at about 14:30, the wifi dropped out, the controller got upset, rebooted, and loaded 37 from the settings.txt. But it was working :-)
![image](https://user-images.githubusercontent.com/104502877/195358825-50ef03c4-d819-44dc-b3a6-9a65f769569d.png)

**Reply by visualapproach** [reply]
Nice, Thank you for testing. I'll add checks for other events now that I know it works.

**Reply by visualapproach** [reply]
The VT should be on par with the 6 wire version now.

**Comment by hubaksis** [comment]
Interesting approach. 

I have a workaround that already works for almost a year.
My spa is connected to OpenHab via MQTT. 
I have two items: actual temperature and display temperature.
Every 30 minutes OpenHab requests actual temperature by turning on the pump for 1 or 2 minutes and then reads the temperature. This is the accurate water temperature.

So, if you want to get the actual temperature for any adjustments/checks, maybe it's possible to add an option to save the temperature only if the pump is on (and for example turn on the pump for 2 minutes every half an hour)?

Cheers.

**Reply by visualapproach** [reply]
Yes, running the pump is the only way to get a good measurement, and scheduling it as you mention is a great way. The virtual temp is filling in the gaps by estimating the temp to fractions of a degree when there is no trustworthy measurements available. Of course the error will increase over time so running the pump every now and then will correct that. In this code the pump must run for 5 min before the measurement is trusted and therefore used to update the virtual temperature.
Cheers
